chainer/functions/normalization/batch_normalization.py

```diff
@@ -15,6 +15,233 @@
 import chainerx
 
 
+class _BatchNormalizationBackend:
+
+    def forward(self, axis, gamma, x, xp, expander,
+                beta, eps, decay, running_mean, running_var):
+        raise NotImplementedError()
+
+    def backward(self, axis, gamma, gy, x, xp,
+                 expander, mean, inv_std, eps, var):
+        raise NotImplementedError()
+
+
+class GeneralBatchNormalizationBackend(_BatchNormalizationBackend):
+
+    def __init__(self):
+        self.mean = None
+        self.inv_std = None
+        self.var = None
+        self.running_mean = None
+        self.running_var = None
+
+    def forward(self, axis, gamma, x, xp, expander,
+                beta, eps, decay, running_mean, running_var):
+        self.running_mean = running_mean
+        self.running_var = running_var
+
+        interm_dtype = numpy.promote_types(x.dtype, gamma.dtype)
+
+        gamma = gamma[expander].astype(interm_dtype, copy=False)
+        beta = beta[expander].astype(interm_dtype, copy=False)
+        self.mean, var = self._get_mean_and_var(axis, gamma,
+                                                x, xp, interm_dtype)
+        if xp is numpy:
+            self.inv_std = numpy.reciprocal(numpy.sqrt(
+                var + eps, dtype=interm_dtype))
+        else:
+            self.inv_std = cuda.cupyx.rsqrt(
+                var + eps, dtype=interm_dtype)
+
+        y = _apply_bn_fwd(xp, x, self.mean[expander],
+                          self.inv_std[expander], gamma, beta)
+
+        # Update running statistics if given
+        if self.running_mean is not None:
+            m = x.size // gamma.size
+            adjust = m / max(m - 1., 1.)  # unbiased estimation
+
+            xp = backend.get_array_module(
+                self.running_mean, self.running_var)
+
+            if xp is chainerx:
+                self.running_mean, self.running_var = backend.from_chx(
+                    (self.running_mean, self.running_var))
+
+            if xp is numpy:
+                self.running_mean *= decay
+                self.running_mean += (1 - decay) * self.mean
+                self.running_var *= decay
+                self.running_var += (1 - decay) * adjust * var
+            else:
+                cuda.elementwise(
+                    'U mean, U var, U decay, U adjust',
+                    'U r_mean, U r_var',
+                    '''
+                    r_mean = r_mean * decay + mean * (1 - decay);
+                    r_var = r_var * decay + var * (1 - decay) * adjust;
+                    ''',
+                    'update_mean_var')(self.mean, var, decay, adjust,
+                                       self.running_mean, self.running_var)
+
+            if xp is chainerx:
+                self.running_mean = backend.to_chx(self.running_mean)
+                self.running_var = backend.to_chx(self.running_var)
+
+        return y,
+
+    def _get_mean_and_var(self, axis, gamma, x, xp, interm_dtype):
```

Comment:
As this method is overridden, it's not internal. Please remove the prefix `_`.

Reply:
Addressed in 4d0e801